components/places/ffi/src/lib.rs

```diff
@@ -273,25 +297,27 @@ pub unsafe extern "C" fn sync15_history_sync(
     error: &mut ExternError,
 ) {
     log::debug!("sync15_history_sync");
-    CONNECTIONS.call_with_result(error, handle, |conn| -> places::Result<_> {
-        // XXX - this is wrong - we kinda want this to be long-lived - the "Db"
-        // should own the store, but it's not part of the db.
-        let store = HistoryStore::new(conn);
-        let mut sync_ping = telemetry::SyncTelemetryPing::new();
-        let result = store.sync(
+    APIS.call_with_result(error, handle, |api| -> places::Result<_> {
+        // Note that api.sync returns a SyncPing which we drop on the floor.
+        // XXX - we should consider exposing it over the FFI while making the
+        // change to PlacesApi.
+        api.sync(
             &sync15::Sync15StorageClientInit {
                 key_id: rust_string_from_c(key_id),
                 access_token: rust_string_from_c(access_token),
                 tokenserver_url: parse_url(rust_str_from_c(tokenserver_url))?,
             },
             &sync15::KeyBundle::from_ksync_base64(rust_str_from_c(sync_key))?,
-            &mut sync_ping,
-        );
-        result
+        )?;
+        Ok(())
     })
 }
 
 define_string_destructor!(places_destroy_string);
 define_bytebuffer_destructor!(places_destroy_bytebuffer);
+define_handle_map_deleter!(APIS, places_api_destroy);
+// XXX - places_connection_destroy should probably be replaced (or maybe just
```

Comment:
Replaced, probably. You can look at the function created by the macro, it's pretty simple.

Note that not returning it via close_connection, and not freeing the PlacesApi from the hashmap mean we never allow you to open a writer after the first one.